### ISSUE TITLE
[SMC] Investigate the Performance Impact Of Current Shader Options

### ISSUE THREAD

**Issue by santorac** [o3de/o3de#11792]
**Is your feature request related to a problem? Please describe.**
We need to optimize the priority order of the shader options in the core material types so that the highest priority, highest impact options appear closest to the root of the shader variant tree. That way, if a requested shader variant is not available and a fallback is selected instead, there is a high chance that the fallback variant will still have baked the most impactful shader options. And in order to determine the optimal order, we need to do some investigation to determine which shader options have the most impact on performance when they are baked vs dynamic.

**Describe the solution you'd like**
Run some manual testing to see if we can figure out which shader options have the biggest performance impact, and document whatever improvement is observed. The general idea is to create a scene that forces lots of overdraw, to tax the main material and lighting shader code, so that any performance gains will be quite obvious. Then systematically change the baking of individual shader options and note the change in performance.

1. Create a test material, using EnhancedPbr.material type, that avoids depth testing. The reason we want EnhancedPbr is that uses the most shader options, including all options that are available in StandardPbr. There are a couple ways to set up the material to avoid depth testing (ideally we would test both, and compare the results).
    1. Set the opacity mode to "Blended". This will disable depth-write, so the entities using it will not cull any other entities. This disadvantage to this approach is the shader won't be exactly the same as the typical forward pass shader, especially around IBL code (see "isTransparent" in Ibl.azsli).
    1. Edit EnhancedPBR_ForwardPass_EDS.shader and EnhancedPBR_ForwardPass.shader to set the depth "WriteMask" to 0, and then  you can test with opacity mode "Opaque" and/or "Cutout" to test the actual forward pass shaders.
1. Set up a .shadervariantlist file with only one variant in it, which exactly matches the test material. Apply the test material to an object in the editor, and use the new material details tool to confirm which variant is being requested. You can even copy the JSON from the material details tool, and paste it directly into the .shadervariantlist file. https://github.com/o3de/o3de/pull/11623 
3. Create a scene with lots of overdraw: Make a new level, assign your test material to the ground plane, and look down at the ground so the plane fills the screen. Duplicate the plane many times (100? 1000?) until the frame time drops significantly, to maybe 1 FPS. 
4. Open the material details tool so you can monitor the changes to the shader variants
5. Note the frame time, this is the optimal frame time since all shader options are baked.
6. Test changing each shader option from baked to dynamic. For each shader option
    1. Edit the .shadervariantlist to delete that shader option from the variant
    2. Wait for the Editor to hot reload the changes. You should see the active shader option value change to -1.
    3. Note the new frame time.
    4. Edit the .shadervariantlist to restore the shader option that you deleted
7. Consider doing a similar test where all shader options are dynamic except for one.

**Describe alternatives you've considered**
Going through shader options one at a time is a good start, and is hopefully enough to give us the results we need. But @antonmic had an idea for a hydra script that will automate a more exhaustive search of the shader option permutation space, running a scenario like what's described above, repeatedly with all kinds of different combinations of shader options baked in. This could be run overnight for multiple nights if necessary, to search for possible combinations of multiple shader options that are significant. This could be done as a separate task if this investigation proves insufficient.

**Additional context**

**Comment by santorac** [comment]
One more thing I recommend: consider adding some new shader options to Lights.azsli for o_enableSimplePointLights, o_enableSimpleSpotLights, o_enablePointLights, o_enableDiskLights, o_enableCapsuleLights, o_enableQuadLights, and o_enablePolygonalLights ... and test these ones too. We are planning to add options like these sometime soon (I believe @invertednormal is working on this) so we can further optimize performance based on the needs of the active scene. As long as you are doing this testing, it would be great if we could get a head start on testing these options as well.

**Comment by santorac** [comment]
Yes, this is on the right track. Here are some thoughts/suggestions...

1. Please report performance in milliseconds instead of fps (you can include FPS too if you want, but I think ms will be more useful).
2. With your test scene, what is the performance difference between a fully baked variant (all options have values at compile time) vs the root variant (all options are dynamic)? That will give us the theoretical maximum improvement, which will be nice to compare with the improvement for individual shader options.
3. The most significant factor for performance should be register pressure. The more registers a shader uses, the fewer threads can be run in parallel on the GPU. By baking shader options, we are reducing register usage. I speculate that when register usage is already low (a fully baked variant), adding few more registers (removing some baked options) won't make much difference. Similarly, when register usage already high (root variant), freeing up just a few registers maybe won't make much difference. So I'm curious how many shader options do you need to add/remove before you start seeing a significant performance difference.

**Comment by dmcdiarmid-ly** [comment]
Radeon GPU Analyzer might be worth investigating for the register analysis portion:
https://github.com/GPUOpen-Tools/radeon_gpu_analyzer

**Comment by moudgils** [comment]
[ATOM-ObtainingRegisterCountinNsight-080922-2056-128.pdf](https://github.com/o3de/o3de/files/9530554/ATOM-ObtainingRegisterCountinNsight-080922-2056-128.pdf)
[ATOM-131656827-080922-2056-126.pdf](https://github.com/o3de/o3de/files/9530555/ATOM-131656827-080922-2056-126.pdf)
PDFs containing info on how to use NSight/Pix to get register usage

**Comment by etherealchain** [comment]
I try to compare the fully baked variant vs the root variant by deleting all options in .shadervariantlist
I confirmed everything is dynamic in material details tool.
Though I didn't see large difference in performance, both of them are around 117~119 ms per frame in my environment.
I am trying NSight now and see if I can get any useful information.

**Comment by santorac** [comment]
Hm, maybe something is wrong with the test procedure. When I was testing with Amazon's "Loft" performance benchmark level, I saw about a 1.5ms difference between root variant and baked variant, compared to about a 30ms total frame time.

**Comment by santorac** [comment]
Did you try adding new flags for the light types? I wonder if that would make more of a difference. 
https://github.com/o3de/o3de/issues/11792#issuecomment-1238936905

**Comment by etherealchain** [comment]
I will try that too.

**Comment by etherealchain** [comment]
I add new options in LightingOptions.azsli and add them in Light.azsli like this
```
if(o_enableSimplePointLights)
{
    ApplySimplePointLights(surface, lightingData);
}
if(o_enableSimpleSpotLights)
{
    ApplySimpleSpotLights(surface, lightingData);
}
if(o_enablePointLights)
{
    ApplyPointLights(surface, lightingData);
}
if(o_enableDiskLights)
{
    ApplyDiskLights(surface, lightingData);
}
if(o_enableCapsuleLights)
{
    ApplyCapsuleLights(surface, lightingData);
}
if(o_enableQuadLights)
{
    ApplyQuadLights(surface, lightingData);
}
if(o_enablePolygonalLights)
{
    ApplyPolygonLights(surface, lightingData);
}
```
Still, there seems to be no difference when on/off these options in .shadervariantlist
One thing I notice is if you leave it running for a while, you can actually see a big difference between full backed and root
full baked is about 90ms and root is about 190ms
test in a scene with 500 plane mesh in blended mode
Editor had been running for at least 30 mins
my card is RTX 2080 Ti

**Comment by etherealchain** [comment]
But it's hard to duplicate this behavior.
At least I only see it one time.

**Comment by etherealchain** [comment]
Update
I'm able to get the register number from NSight, though I'm not sure if it make sense

This is from one DrawIndexedInstanced call
Option list
```
"o_applySpecularAA": "false",
"o_area_light_validation": "false",
"o_baseColorTextureBlendMode": "TextureBlendMode::Multiply",
"o_baseColor_useTexture": "true",
"o_blend_between_cascades_enable": "false",
"o_clearCoat_enabled": "false",
"o_clearCoat_factor_useTexture": "false",
"o_clearCoat_feature_enabled": "false",
"o_clearCoat_normal_useTexture": "false",
"o_clearCoat_roughness_useTexture": "false",
"o_detail_baseColor_useTexture": "false",
"o_detail_blendMask_useTexture": "false",
"o_detail_normal_useTexture": "false",
"o_diffuseOcclusion_useTexture": "false",
"o_directional_shadow_filtering_method": "ShadowFilterMethod::Pcf",
"o_directional_shadow_receiver_plane_bias_enable": "true",
"o_emissiveEnabled": "false",
"o_emissive_useTexture": "false",
"o_enableAnisotropy": "false",
"o_enableAreaLights": "true",
"o_enableDirectionalLights": "true",
"o_enableIBL": "true",
"o_enablePunctualLights": "true",
"o_enableShadows": "true",
"o_enableSubsurfaceScattering": "false",
"o_materialUseForwardPassIBLSpecular": "false",
"o_meshUseForwardPassIBLSpecular": "false",
"o_metallic_useTexture": "false",
"o_normal_useTexture": "false",
"o_opacity_mode": "OpacityMode::Blended",
"o_opacity_source": "OpacitySource::Packed",
"o_parallax_algorithm": "ParallaxAlgorithm::POM",
"o_parallax_enablePixelDepthOffset": "false",
"o_parallax_feature_enabled": "false",
"o_parallax_highlightClipping": "false",
"o_parallax_quality": "ParallaxQuality::Low",
"o_parallax_shadow": "false",
"o_roughness_useTexture": "false",
"o_shader_debugging_enabled": "false",
"o_specularF0_enableMultiScatterCompensation": "false",
"o_specularF0_useTexture": "false",
"o_specularOcclusion_useTexture": "false",
"o_subsurfaceScattering_useTexture": "false",
"o_transmission_mode": "TransmissionMode::None",
"o_transmission_useTexture": "false",
"o_useHeightmap": "false"
```

Fully backed
![image](https://user-images.githubusercontent.com/4178768/190129170-a8bccbf3-149c-4ed5-958d-b6baf946c663.png)

Root variant
![image](https://user-images.githubusercontent.com/4178768/190129799-de216ad1-b8be-4a64-ad44-1c72775bcd2a.png)

**Comment by jeremyong-az** [comment]
Per discussion yesterday, I would suggest as another test (possibly simpler) simply having a single draw in the scene filling the viewport with a single material. This material then should swap to the different variations being tested including the root variant (it may be useful to make a cvar or imgui menu to control this, or perhaps a simple game script that lets you cycle through variations using arrow keys or something). Then, you should be able to capture timings for that specific draw in a tool like Nsight (I suggest doing a fullscreen capture optionally).

**Comment by santorac** [comment]
Yeah, once you have it narrowed down to a single plane, we should start with a comparison of the register count for root variant vs fully baked variant. We should expect a significant difference there.

Then the next step would be to test individual shader options, baked vs dynamic, to see how much each one impacts register count. Some might be interdependent though, like multiple shader options that would need to all be baked together to reduce register count.

It would also be interesting to see the difference between all light types enabled, vs all light types disabled and baked out, vs only area lights disabled and baked out.

**Comment by siliconvoodoo** [comment]
Maybe the 2080ti is too powerful to manifest performance discrepancies by variant, because of too fast ALU. There's got to be a bottleneck in the ROPs. Or, inversely, the size(code) is >>> wrt size(flow-graph-code) which hides the cost gained by branchless. The options are constant (uniforms) so branch hardware is 100% predictive isn't it; VGPR pressure is the only thing that theoretically can have an impact, but does wave occupancy matter that much on such a high end card, say if memory accesses hides the occupancy?

**Comment by etherealchain** [comment]
Update
I did a simple experiment to test the impact of shader variant on registers
Here is the full list
![image](https://user-images.githubusercontent.com/4178768/192258831-7e23c63c-f213-4ae0-83cb-8df91118221e.png)

With full baked the PS register is 128, and root variant is 168
First I test on delete variant one at a time, but there is not much difference on register number, 
usually they stay the same, sometimes even become smaller

Next I test on delete a bunch of them, and find out there is a some what unclear threshold
if there are enough variant baked, the register number will be 128, otherwise it's 168
I didn't know much about the relation between variants
So I just test on delete from top and bottom

As you can see in the picture
Delete from top ( variant in red box are baked) and delete from bottom (variant in blue box are backed) are very different

I am going to test on https://github.com/o3de/o3de/issues/11792#issuecomment-1238936905 next to see if there is any difference

**Comment by santorac** [comment]
That's really interesting. I would do some extra testing around the thresholds where you saw the register counts jump considerably. I wonder if there is significance to those particular registers. In the blue box, the threshold is around directional light shadow options, which is not surprising, I think remember someone else saying they noticed shadow filtering having a significant impact on register count. And in the red box, the threshold is near enableDirectionalLights and enableShadows, it might not be a coincidence. So I put some more focus on directional_shadow_*, enableDirectionalLights, and enableShadows.

**Comment by santorac** [comment]
I would also recommend doing a test to make sure that code optimization is working the way we want. The variant system assumes that DXC's dead-code-elimination will remove the code for baked shader options. But if this assumption is wrong for some reason, and DXC isn't doing all the optimization we expect, this could cripple the benefit of the variant system.

Here is how I would test that: Start with your fully-baked variant, and the PS register count of 128. Now locally edit the shader code to remove every place where the code references a shader option. For example, if your baked shader option is o_directional_shadow_filtering_method=Pcf, then go _delete_ these functions and the related code: GetVisibilityFromLightNoFilter(), GetVisibilityFromLightEsm(), GetVisibilityFromLightEsmPcf(). Similarly, since o_parallax_feature_enabled=false, _delete_ all parallax code. After doing this for a few shader options, check the PS register count again. If it starts going down below 128, then we have some bigger problems to address.

**Comment by jeremyong-az** [comment]
@etherealchain I have a few suggestions to help your workflow here.

First, I suggest downloading a recent RGA release (link: https://gpuopen.com/rga/#download). Installed with RGA is an `rga.exe` executable which may be used to compile shaders directly to AMD's ISA (you specify the asic), which can then be used to determine the register count. Note that an AMD card and driver are _not needed_ to do this compilation (there is an `--offline-*` compilation mode to use).

Getting the full register count for a compiled pixel shader or vertex shader requires that the shader be compiled in the context of a full graphics pipeline, meaning you need to specify pipeline state and both vertex and entry points. Once you set it up once though, you'll be able to very quickly make changes and see the output without needing to continuously do frame captures. Also, this will be a workflow we can enable for all engineers (since an AMD card isn't needed, unlike the Nvidia workflow).

To see what the RGA output sort of looks like, here's a CE example: https://godbolt.org/z/Ea3dKKjvj

**Comment by etherealchain** [comment]
> I would also recommend doing a test to make sure that code optimization is working the way we want. The variant system assumes that DXC's dead-code-elimination will remove the code for baked shader options. But if this assumption is wrong for some reason, and DXC isn't doing all the optimization we expect, this could cripple the benefit of the variant system.
> 
> Here is how I would test that: Start with your fully-baked variant, and the PS register count of 128. Now locally edit the shader code to remove every place where the code references a shader option. For example, if your baked shader option is o_directional_shadow_filtering_method=Pcf, then go _delete_ these functions and the related code: GetVisibilityFromLightNoFilter(), GetVisibilityFromLightEsm(), GetVisibilityFromLightEsmPcf(). Similarly, since o_parallax_feature_enabled=false, _delete_ all parallax code. After doing this for a few shader options, check the PS register count again. If it starts going down below 128, then we have some bigger problems to address, it means the shader variant system doesn't work as intended.


I think I can confirem dead-code-elimination is working correctly by following test:
This form shows the result of vgpr with different lighting option and material setting
The test is done with shader variant list fully baked
image

And when I manually remove code in ApplyDirectLighting, I also get a vgpr count = 57

**Comment by etherealchain** [comment]
But in the case of o_directional_shadow_filtering_method, I did get a smaller vgpr count = 124, when I remove GetVisibilityFromLightNoFilter(), GetVisibilityFromLightEsm() and GetVisibilityFromLightEsmPcf() manually.

**Comment by etherealchain** [comment]
In the case of parallax mapping, it's more difficult to find all the code that should be deleted, but I havn't get a smaller vpgr value so far.

**Comment by santorac** [comment]
@rgba16f did some investigation into register counts a while back. He was testing deleting parts of the shader code, rather than baking variants, and checked register counts. Here is what he reported seeing:
![image](https://user-images.githubusercontent.com/55155825/193142460-da59f735-88b3-489c-8392-66342309464f.png)

He also noted that he dug further into Direct Lighting and shadow visibility testing is one of the main sources of register usage. For directional lights it accounted for 40 vgpr's, for projected shadows it accounted for 22 vgpr's.

**Comment by santorac** [comment]
I like the testing you did above with turning off some features in the material and comparing baked vs not-baked. Most of the other tests so far have been using more of a default StandardPBR material, with a fair number of features enabled. Going forward, I suggest focusing on testing a material with as _many features turned off_ as possible in the material data. That will give us better insight into the register impact when you bake shader options.

Definitely try to get RGA set up first though, as @jeremyong-az suggested, so you don't have to do runtime captures and this will cut down your iteration time.

Hopefully then we can get a good picture of which shader options are most important.

**Comment by etherealchain** [comment]
Update
I try to setup the pipeline state and get Spir-V binary using `dxc -spirv -T vs_6_2 -E VertexShader D:\o3de-project\myproject\Cache\pc\shaders\materials\enhancedpbr\enhancedpbr_forwardpass_eds_vulkan.hlsl`

but so far I am unable to build with rga, even with a simple command like
`rga.exe -s vk-spv-offline -c gfx1035 --isa "D:\output\isa.txt" --il "D:\output\il.txt" -a "D:\output\stats.csv" --vert "D:\vertex_vulkan.spv" --frag "D:\pixel_vulkan.spv"`

it always report the same error
```
Building for gfx1035... failed.
Error: Options: -Dall -l -gfxip 1035 -set in.spv in.spv out.comp.palIlText out.vert.palIlText out.tesc.palIlText out.tese.palIlText out.geom.palIlText out.frag.palIlText out.comp.isa out.vert.isa out.tesc.isa out.tese.isa out.geom.isa out.frag.isa out.comp.isaText out.vert.isaText out.tesc.isaText out.tese.isaText out.geom.isaText out.frag.isaText out.comp.isaInfo out.vert.isaInfo out.tesc.isaInfo out.tese.isaInfo out.geom.isaInfo out.frag.isaInfo out.glslLog defaultOutput -set val
AMD-SC: Error: Assertion failed: !options.last_pass() || color.IsValid() || ( blk->IsIfFooter() && m_whole_wave_ranges->get( ii ) ) at C:\Jenkins\workspace\vulkan\vulkan-psdb\driver_build\drivers\xgl\icd\imported\sc\Src\NSrc\SCRegAlloc.cpp(7463)
Error: SC internal error.
Fail to link graphics pipeline.
```

I have tried uing the gui and the command, tried different gpu target, tried different shader model, but I always get the same error.
I always tried `enhancedpbr_forwardpass_eds_dx12.hlsl` and `enhancedpbr_forwardpass_eds_null.hlsl`, but they report a different error 
```
Building for gfx1035... failed.
Error: Error: The specified AMDIL is invalid or corrupt.
Fail to link graphics pipeline.
```

**Comment by etherealchain** [comment]
Update

I have switched to Radeon RX Vega, however both live and offline  mode are still failing.
So I test through all supported Target GPUs `gfx1035, gfx1034, gfx1032, gfx1031, gfx1030, gfx1012, gfx1011, gfx1010, gfx90c, gfx906, gfx902, gfx900`, all gfx10XX failed both live and offline mode, with the same error, all gfx90X failed live mode, but are able to build on offline mode.
I have not test gfx90X on Nvidia card, so I'm not sure if the successful build is caused by amd card.

Anyway, after the successful build, you can see a VGPR pressure visualization on the right side.
The fragment shader result is quite large though, not sure how to do a reasonable comparison between different sahder option

![image](https://user-images.githubusercontent.com/4178768/193778958-69acaa7c-2ba2-469e-8a22-24089f34a1e8.png)

**Comment by etherealchain** [comment]
Update 
I am able to get 2 disassembly results from 2 different light setting
The settings are the same as above (light option and material both on & light option and material both off)

The command I use to get spirv
`dxc.exe -spirv -T ps_6_0 -E PixelShader D:\o3de-project\myproject\user\AssetProcessorTemp\JobTemp-hnMupC\EnhancedPBR_ForwardPass_EDS_vulkan_1.hlsl -Fo D:\pixel_vulkan.spv`

The rga command I use
`rga.exe -s vk-spv-offline --isa "D:\RgaOutput\disassem.txt" --parse-isa --livereg "D:\RgaOutput\livereg.txt" --line-numbers --analysis "D:\RgaOutput\resourceUsage.csv" -b "D:\RgaOutput\codeobj.bin" --session-metadata "D:\RgaOutput\gfx90c_cliInvocation.xml" --asic gfx90c "D:\pixel_vulkan.spv"`

The results are attached here
[LightOff.zip](https://github.com/o3de/o3de/files/9731953/LightOff.zip)
[LightOn.zip](https://github.com/o3de/o3de/files/9731954/LightOn.zip)

**Comment by etherealchain** [comment]
Update

I did some more tests, so far lighting still has the most impact (using LightOn.zip as reference).
All tests have to update both shader variant list and material file so we can get a match

[ClearCoatOn.zip](https://github.com/o3de/o3de/files/9754884/ClearCoatOn.zip)
[ParallaxOn.zip](https://github.com/o3de/o3de/files/9754885/ParallaxOn.zip)
[ShadowOff.zip](https://github.com/o3de/o3de/files/9754886/ShadowOff.zip)

Most of them make sense except parallax, enable parallax doesn't change register number on my pc.

**Comment by santorac** [comment]
We have determined that the best way forward is to integrate RGA into O3DE for convenient offline analysis of these shader options, rather than attempting manual processes. Here is a new Issue to track that work: https://github.com/o3de/o3de/issues/12550
